### DISCUSSION TITLE
Generate `useAsyncData` key instead of explicitly specifying it

### DISCUSSION THREAD

**Discussion by mvrlin** [nuxt/nuxt#17056] [Ideas]
I propose to generate the key on the Nuxt side, instead of specifying it explicitly when calling the function, to solve this problem you can generate the hash from the body of the function.

Here's the example:

```typescript
// composables/asyncData.ts

function generateHash(input: string) {
  let seed: number | string = 0

  for (let i = 0; i < input.length; i++) {
    seed = (seed << 5) - seed + input.charCodeAt(i)
    // Convert to 32bit integer
    seed |= 0
  }

  return seed.toString(16)
}

export function useAsyncData<
  DataT,
  Transform extends _Transform<DataT> = _Transform<DataT, DataT>,
  PickKeys extends KeyOfRes<Transform> = KeyOfRes<Transform>
> (
  handler: (ctx?: NuxtApp) => Promise<DataT>,
  options: AsyncDataOptions<DataT, Transform, PickKeys> = {}
) : AsyncData<PickFrom<ReturnType<Transform>, PickKeys>> {
  if (typeof handler !== 'function') {
    throw new TypeError('asyncData handler must be a function')
  }
  
  const key = generateHash(handler.toString())
  
  // ...
```

Before:
```vue
<script lang="ts" setup>
const { data } = await useAsyncData('example', () => $fetch('api.example.com'))
</script>
```

After:
```vue
<script lang="ts" setup>
const { data } = await useAsyncData(() => $fetch('api.example.com'))
</script>
```

**Comment by pi0** [comment]
Thanks for the idea! Actually, `useFetch` which is based on `useAsyncData` does this based on request. 

```vue
<script lang="ts" setup>
// Key is auto generated like $f_api_example_com
const { data } = await useFetch('api.example.com')
</script>
```

We taught of doing it for `useAsyncData` too but generating keys based on serialized function can be tricky as the getter body might be the same but depending on external variables changing behavior. This way we would wrongly make conflicts. 

Other alternatives:
- A transform step similar to `ssrRef` in `@nuxtjs/composition-api` might work but needs extensive AST parsing. 
- Based on order similar to fetch keys of nuxt: Is unstable

I think using explicit key makes the best sense for longer-term and more composables similar to `useFetch` can be built on top of a simple unit like `useAsyncData` to offer smart and stable key generation.

**Reply by mvrlin** [reply]
Oh yea, that makes sense 😅

**Comment by ol1s** [comment]
I'm interested - where does `useFetch()` get it's `autoKey` from? I'm looking at the source code and it looks like magic!!

```
export function useFetch(request, arg1, arg2) {
  const [opts, autoKey] = typeof arg1 === "string" ? [{}, arg1] : [arg1, arg2];
  //etc
}
```
